frappe/modules/import_file.py

```diff
@@ -97,8 +98,15 @@ def import_doc(docdict, force=False, data_import=False, pre_process=None,
 		ignore_version=None, reset_permissions=False):
 	frappe.flags.in_import = True
 	docdict["__islocal"] = 1
+
+	controller = get_controller(docdict['doctype'])
+	if controller and hasattr(controller, 'prepare_docdict_for_import') and callable(getattr(controller, 'prepare_docdict_for_import')):
```

Comment:
hardcoded, use generic (`before_import`)

Reply:
Can't use `before_import` for this case! Reason is that `frappe.get_doc(docdict)` fails when there is a list attribute: `field_order` will cause frappe to throw in `frappe.get_doc()` (https://github.com/frappe/frappe/blob/develop/frappe/model/base_document.py#L155-L162).

Since I can't get the `Document` from the `docdict`, I can't run the `before_import` method. And to run `frappe.get_doc()` I'm preprocessing the `docdict` in the staticmethod `prepare_docdict_for_import` and removing the `field_order` list.

`before_import` is a valid event now though.

Reply:
That is why i was thinking we can have `[doctype]_layout.json` as a separate file 😉 

Just change the fieldname to `prepare_for_import`, reads a bit cleaner and will merge it.. 👍

Reply:
I thought about having a separate `[doctype]_layout.json` too but I think we would face a few issues with with:
* There will be multiple timestamps (to check before migrating)
* Discrepancy between the fieldnames in layout and and doctype JSON could arise and would be more difficult to catch than with just a single file

So I'll rename the function name to `prepare_for_import`